semi reports missing semicolon on export default class

```js
// test.js
export default class {};
```

```js
// config.json
{
	"env": {
	  "es6": true
	},
	"ecmaFeatures": {
	  "modules": true
	},
    "rules": {
        "semi": 2
    }
}
```

```sh
$ eslint --reset --no-eslintrc --config config.json test.js

test.js
  2:23  error  Missing semicolon  semi

✖ 1 problem (1 error, 0 warnings)
```

Looks like similar cases were addressed in #2178 and this one was missed.

--- a/lib/rules/semi.js
+++ b/lib/rules/semi.js
@@ -9,7 +9,7 @@
 //------------------------------------------------------------------------------
 module.exports = function(context) {
 
-    var OPT_OUT_PATTERN = /[\[\(\/\+\-]/;
+    var OPT_OUT_PATTERN = /[\[\(\/\+\-]/; // One of [(/+-
 
     var always = context.options[0] !== "never";
 
@@ -18,19 +18,48 @@ module.exports = function(context) {
     //--------------------------------------------------------------------------
 
     /**
+     * Reports a semicolon error with appropriate location and message.
+     * @param {ASTNode} node The node with an extra or missing semicolon.
+     * @returns {void}
+     */
+    function report(node) {
+        var message = always ? "Missing semicolon." : "Extra semicolon.";
+        context.report(node, context.getLastToken(node).loc.end, message);
+    }
+
+    /**
+     * Checks whether a token is a semicolon punctuator.
+     * @param {Token} token The token.
+     * @returns {boolean} True if token is a semicolon punctuator.
+     */
+    function isSemicolon(token) {
+        return (token.type === "Punctuator" && token.value === ";");
+    }
+
+    /**
      * Check if a semicolon is unnecessary, only true if:
      *   - next token is on a new line and is not one of the opt-out tokens
      *   - next token is a valid statement divider
      * @param {Token} lastToken last token of current node.
-     * @param {Token} nextToken next token after current node.
      * @returns {boolean} whether the semicolon is unnecessary.
      */
-    function isUnnecessarySemicolon(lastToken, nextToken) {
+    function isUnnecessarySemicolon(lastToken) {
+        var isDivider, isOptOutToken, lastTokenLine, nextToken, nextTokenLine;
 
-        var lastTokenLine = lastToken.loc.end.line,
-            nextTokenLine = nextToken && nextToken.loc.start.line,
-            isOptOutToken = nextToken && OPT_OUT_PATTERN.test(nextToken.value),
-            isDivider = nextToken && (nextToken.value === "}" || nextToken.value === ";");
+        if (!isSemicolon(lastToken)) {
+            return false;
+        }
+
+        nextToken = context.getTokenAfter(lastToken);
+
+        if (!nextToken) {
+            return true;
+        }
+
+        lastTokenLine = lastToken.loc.end.line;
+        nextTokenLine = nextToken.loc.start.line;
+        isOptOutToken = OPT_OUT_PATTERN.test(nextToken.value);
+        isDivider = (nextToken.value === "}" || nextToken.value === ";");
 
         return (lastTokenLine !== nextTokenLine && !isOptOutToken) || isDivider;
     }
@@ -41,19 +70,15 @@ module.exports = function(context) {
      * @returns {void}
      */
     function checkForSemicolon(node) {
-        var lastToken = context.getLastToken(node),
-            nextToken = context.getTokenAfter(node);
+        var lastToken = context.getLastToken(node);
 
         if (always) {
-            if (lastToken.type !== "Punctuator" || lastToken.value !== ";") {
-                context.report(node, lastToken.loc.end, "Missing semicolon.");
+            if (!isSemicolon(lastToken)) {
+                report(node);
             }
         } else {
-            if (lastToken.type === "Punctuator" &&
-                lastToken.value === ";" &&
-                isUnnecessarySemicolon(lastToken, nextToken)
-            ) {
-                context.report(node, node.loc.end, "Extra semicolon.");
+            if (isUnnecessarySemicolon(lastToken)) {
+                report(node);
             }
         }
     }
@@ -64,7 +89,6 @@ module.exports = function(context) {
      * @returns {void}
      */
     function checkForSemicolonForVariableDeclaration(node) {
-
         var ancestors = context.getAncestors(),
             parentIndex = ancestors.length - 1,
             parent = ancestors[parentIndex];
@@ -91,21 +115,26 @@ module.exports = function(context) {
         "ContinueStatement": checkForSemicolon,
         "ImportDeclaration": checkForSemicolon,
         "ExportAllDeclaration": checkForSemicolon,
-        "ExportNamedDeclaration": checkForSemicolon,
-        "ExportDefaultDeclaration": checkForSemicolon,
+        "ExportNamedDeclaration": function (node) {
+            if (!node.declaration) {
+                checkForSemicolon(node);
+            }
+        },
+        "ExportDefaultDeclaration": function (node) {
+            if (!/(?:Class|Function)Declaration/.test(node.declaration.type)) {
+                checkForSemicolon(node);
+            }
+        },
         "EmptyStatement": function (node) {
-            var lastToken, nextToken;
+            var lastToken;
 
             if (!always) {
                 lastToken = context.getLastToken(node);
-                nextToken = context.getTokenAfter(node) || context.getLastToken(node);
 
-                if (isUnnecessarySemicolon(lastToken, nextToken)) {
-                    context.report(node, "Extra semicolon.");
+                if (isUnnecessarySemicolon(lastToken)) {
+                    report(node);
                 }
             }
-
-
         }
     };
 

--- a/tests/lib/rules/semi.js
+++ b/tests/lib/rules/semi.js
@@ -41,10 +41,46 @@ eslintTester.addRuleTest("lib/rules/semi", {
         { code: "a++\nb++", options: ["never"] },
         { code: "a++; b++", options: ["never"] },
         { code: "for (let thing of {}) {\n  console.log(thing);\n}", ecmaFeatures: { forOf: true, blockBindings: true }},
-        { code: "import theDefault, { named1, named2 } from 'src/mylib'", options: ["never"], ecmaFeatures: { modules: true }},
-        { code: "export {foo, bar}", options: ["never"], ecmaFeatures: { modules: true }},
+
         { code: "import theDefault, { named1, named2 } from 'src/mylib';", ecmaFeatures: { modules: true }},
-        { code: "export {foo, bar};", ecmaFeatures: { modules: true }}
+        { code: "import theDefault, { named1, named2 } from 'src/mylib'", options: ["never"], ecmaFeatures: { modules: true }},
+
+        // exports, "always"
+        { code: "export * from 'foo';", ecmaFeatures: { modules: true } },
+        { code: "export { foo } from 'foo';", ecmaFeatures: { modules: true } },
+        { code: "export { foo };", ecmaFeatures: { modules: true } },
+        { code: "export var foo;", ecmaFeatures: { modules: true } },
+        { code: "export function foo () { }", ecmaFeatures: { modules: true } },
+        { code: "export function* foo () { }", ecmaFeatures: { generators: true, modules: true } },
+        { code: "export class Foo { }", ecmaFeatures: { classes: true, modules: true } },
+        { code: "export let foo;", ecmaFeatures: { blockBindings: true, modules: true } },
+        { code: "export const FOO = 42;", ecmaFeatures: { blockBindings: true, modules: true } },
+        { code: "export default function () { }", ecmaFeatures: { modules: true } },
+        { code: "export default function* () { }", ecmaFeatures: { generators: true, modules: true } },
+        { code: "export default class { }", ecmaFeatures: { classes: true, modules: true } },
+        { code: "export default foo || bar;", ecmaFeatures: { modules: true } },
+        { code: "export default (foo) => foo.bar();", ecmaFeatures: { arrowFunctions: true, modules: true } },
+        { code: "export default foo = 42;", ecmaFeatures: { modules: true } },
+        { code: "export default foo += 42;", ecmaFeatures: { modules: true } },
+
+        // exports, "never"
+        { code: "export * from 'foo'", options: ["never"], ecmaFeatures: { modules: true } },
+        { code: "export { foo } from 'foo'", options: ["never"], ecmaFeatures: { modules: true } },
+        { code: "export { foo }", options: ["never"], ecmaFeatures: { modules: true } },
+        { code: "export var foo", options: ["never"], ecmaFeatures: { modules: true } },
+        { code: "export function foo () { }", options: ["never"], ecmaFeatures: { modules: true } },
+        { code: "export function* foo () { }", options: ["never"], ecmaFeatures: { generators: true, modules: true } },
+        { code: "export class Foo { }", options: ["never"], ecmaFeatures: { classes: true, modules: true } },
+        { code: "export let foo", options: ["never"], ecmaFeatures: { blockBindings: true, modules: true } },
+        { code: "export const FOO = 42", options: ["never"], ecmaFeatures: { blockBindings: true, modules: true } },
+        { code: "export default function () { }", options: ["never"], ecmaFeatures: { modules: true } },
+        { code: "export default function* () { }", options: ["never"], ecmaFeatures: { generators: true, modules: true } },
+        { code: "export default class { }", options: ["never"], ecmaFeatures: { classes: true, modules: true } },
+        { code: "export default foo || bar", options: ["never"], ecmaFeatures: { modules: true } },
+        { code: "export default (foo) => foo.bar()", options: ["never"], ecmaFeatures: { arrowFunctions: true, modules: true } },
+        { code: "export default foo = 42", options: ["never"], ecmaFeatures: { modules: true } },
+        { code: "export default foo += 42", options: ["never"], ecmaFeatures: { modules: true } }
+
     ],
     invalid: [
         { code: "import * as utils from './utils'", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ImportDeclaration"}] },
@@ -52,11 +88,6 @@ eslintTester.addRuleTest("lib/rules/semi", {
         { code: "import { default as foo } from 'lib'", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ImportDeclaration"}] },
         { code: "import 'src/mylib'", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ImportDeclaration"}] },
         { code: "import theDefault, { named1, named2 } from 'src/mylib'", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ImportDeclaration"}] },
-        { code: "export {foo, bar}", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportNamedDeclaration"}] },
-        { code: "export {foo} from 'mod'", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportNamedDeclaration"}] },
-        { code: "export default function () {}", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportDefaultDeclaration"}] },
-        { code: "export default 1", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportDefaultDeclaration"}] },
-        { code: "export * from 'mod'", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportAllDeclaration"}] },
         { code: "function foo() { return [] }", errors: [{ message: "Missing semicolon.", type: "ReturnStatement"}] },
         { code: "while(true) { break }", errors: [{ message: "Missing semicolon.", type: "BreakStatement"}] },
         { code: "while(true) { continue }", errors: [{ message: "Missing semicolon.", type: "ContinueStatement"}] },
@@ -92,6 +123,29 @@ eslintTester.addRuleTest("lib/rules/semi", {
         { code: ";", options: ["never"], errors: [{ message: "Extra semicolon.", type: "EmptyStatement"}] },
         { code: ";;", options: ["never"], errors: [{ message: "Extra semicolon.", type: "EmptyStatement"}, { message: "Extra semicolon.", type: "EmptyStatement"}] },
         { code: "import theDefault, { named1, named2 } from 'src/mylib';", options: ["never"], ecmaFeatures: { modules: true }, errors: [{ message: "Extra semicolon.", type: "ImportDeclaration"}] },
-        { code: "export {foo, bar};", options: ["never"], ecmaFeatures: { modules: true }, errors: [{ message: "Extra semicolon.", type: "ExportNamedDeclaration"}] }
+
+        // exports, "always"
+        { code: "export * from 'foo'", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportAllDeclaration" }] },
+        { code: "export { foo } from 'foo'", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportNamedDeclaration" }] },
+        { code: "export { foo }", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportNamedDeclaration" }] },
+        { code: "export var foo", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "VariableDeclaration" }] },
+        { code: "export let foo", ecmaFeatures: { blockBindings: true, modules: true }, errors: [{ message: "Missing semicolon.", type: "VariableDeclaration" }] },
+        { code: "export const FOO = 42", ecmaFeatures: { blockBindings: true, modules: true }, errors: [{ message: "Missing semicolon.", type: "VariableDeclaration" }] },
+        { code: "export default foo || bar", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportDefaultDeclaration" }] },
+        { code: "export default (foo) => foo.bar()", ecmaFeatures: { arrowFunctions: true, modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportDefaultDeclaration" }] },
+        { code: "export default foo = 42", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportDefaultDeclaration" }] },
+        { code: "export default foo += 42", ecmaFeatures: { modules: true }, errors: [{ message: "Missing semicolon.", type: "ExportDefaultDeclaration" }] },
+
+        // exports, "never"
+        { code: "export * from 'foo';", options: ["never"], ecmaFeatures: { modules: true }, errors: [{ message: "Extra semicolon.", type: "ExportAllDeclaration" }] },
+        { code: "export { foo } from 'foo';", options: ["never"], ecmaFeatures: { modules: true }, errors: [{ message: "Extra semicolon.", type: "ExportNamedDeclaration" }] },
+        { code: "export { foo };", options: ["never"], ecmaFeatures: { modules: true }, errors: [{ message: "Extra semicolon.", type: "ExportNamedDeclaration" }] },
+        { code: "export var foo;", options: ["never"], ecmaFeatures: { modules: true }, errors: [{ message: "Extra semicolon.", type: "VariableDeclaration" }] },
+        { code: "export let foo;", options: ["never"], ecmaFeatures: { blockBindings: true, modules: true }, errors: [{ message: "Extra semicolon.", type: "VariableDeclaration" }] },
+        { code: "export const FOO = 42;", options: ["never"], ecmaFeatures: { blockBindings: true, modules: true }, errors: [{ message: "Extra semicolon.", type: "VariableDeclaration" }] },
+        { code: "export default foo || bar;", options: ["never"], ecmaFeatures: { modules: true }, errors: [{ message: "Extra semicolon.", type: "ExportDefaultDeclaration" }] },
+        { code: "export default (foo) => foo.bar();", options: ["never"], ecmaFeatures: { arrowFunctions: true, modules: true }, errors: [{ message: "Extra semicolon.", type: "ExportDefaultDeclaration" }] },
+        { code: "export default foo = 42;", options: ["never"], ecmaFeatures: { modules: true }, errors: [{ message: "Extra semicolon.", type: "ExportDefaultDeclaration" }] },
+        { code: "export default foo += 42;", options: ["never"], ecmaFeatures: { modules: true }, errors: [{ message: "Extra semicolon.", type: "ExportDefaultDeclaration" }] }
     ]
 });

